### PR TITLE
Getting the test cases to pass by fixing up the parsestr unicode problem

### PR DESCRIPTION
**Pull request by coleb**
This problem is holding up an ongoing release we're working on that
will be the debut release on binstar.org for our packages. Please
merge quickly.

Would you be interested in me creating a travis CI for binstar? The test cases would have caught the above problem.

### CODE REVIEW

**Comment by srossross** [comment]
The automated tests are passing for python 2 and 3. https://binstar.org/binstar/binstar/builds

@coleb Can you show me the failing tests?

**Comment by coleb** [comment]
(binstar_client_devel)[11:47][((no branch))]$ which python
/Users/coleb/.virtualenvs/binstar_client_devel/bin/python
(binstar_client_devel)[11:47][((no branch))]$ python --version
Python 2.7.2
(binstar_client_devel)[11:47][((no branch))]$ uname -a
Darwin server-2.local 12.4.0 Darwin Kernel Version 12.4.0: Wed May  1 17:57:12 PDT 2013; root:xnu-2050.24.15~1/RELEASE_X86_64 x86_64
(binstar_client_devel)[11:46][((no branch))]$ git describe 
0.10.3
(binstar_client_devel)[11:46][((no branch))]$ python binstar_client/tests/
.coverage            coverage_report.py   fixture.py           runner.py            test_ipynb.py        test_login.pyc       test_upload.py       test_whoami.pyc      
__init__.py          coverage_report.pyc  fixture.pyc          runner.pyc           test_ipynb.pyc       test_register.py     test_upload.pyc      urlmock.py           
__init__.pyc         data/                requirements.txt     runtests.py          test_login.py        test_register.pyc    test_whoami.py       urlmock.pyc          
(binstar_client_devel)[11:46][((no branch))]$ python binstar_client/tests/runtests.py 
<module 'binstar_client' from '/Users/coleb/binstar/binstar_client/__init__.pyc'>
Discover /Users/coleb/binstar/
test_conda (binstar_client.inspect_package.tests.test_conda.Test) ... ok
test_bdist_egg (binstar_client.inspect_package.tests.test_pypi.Test) ... ERROR
test_bdist_wheel (binstar_client.inspect_package.tests.test_pypi.Test) ... ok
test_sdist (binstar_client.inspect_package.tests.test_pypi.Test) ... ERROR
test_sdist_distutils (binstar_client.inspect_package.tests.test_pypi.Test) ... ERROR
test_inspect_ipynb_2_package (binstar_client.tests.test_ipynb.Test) ... ok
test_inspect_ipynb_3_package (binstar_client.tests.test_ipynb.Test) ... ok
test_login (binstar_client.tests.test_login.Test) ... Using binstar api site https://api.binstar.org
test_user's ok
test_register_private (binstar_client.tests.test_register.Test) ... skipped u'Need to change this to binsar package --create'
test_register_public (binstar_client.tests.test_register.Test) ... skipped u'Need to change this to binsar package --create'
test_upload (binstar_client.tests.test_upload.Test) ... Using binstar api site https://api.binstar.org
ok
test_upload_bad_package (binstar_client.tests.test_upload.Test) ... Using binstar api site https://api.binstar.org
ok
test_upload_bad_package_no_register (binstar_client.tests.test_upload.Test) ... Using binstar api site https://api.binstar.org
ok
test_whoami (binstar_client.tests.test_whoami.Test) ... Using binstar api site https://api.binstar.org
ok
test_whoami_anon (binstar_client.tests.test_whoami.Test) ... Using binstar api site https://api.binstar.org
ok

======================================================================
ERROR: test_bdist_egg (binstar_client.inspect_package.tests.test_pypi.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/coleb/binstar/binstar_client/inspect_package/tests/test_pypi.py", line 101, in test_bdist_egg
    package_data, version_data, file_data = pypi.inspect_pypi_package(filename, fd)
  File "/Users/coleb/binstar/binstar_client/inspect_package/pypi.py", line 369, in inspect_pypi_package
    return inspect_pypi_package_egg(filename, fileobj)
  File "/Users/coleb/binstar/binstar_client/inspect_package/pypi.py", line 275, in inspect_pypi_package_egg
    package_data = {'name': pop_key(attrs, 'Name'),
  File "/Users/coleb/binstar/binstar_client/inspect_package/uitls.py", line 41, in pop_key
    value = data.pop(k, *d)
KeyError: u'Name'

======================================================================
ERROR: test_sdist (binstar_client.inspect_package.tests.test_pypi.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/coleb/binstar/binstar_client/inspect_package/tests/test_pypi.py", line 61, in test_sdist
    package_data, version_data, file_data = pypi.inspect_pypi_package(filename, fd)
  File "/Users/coleb/binstar/binstar_client/inspect_package/pypi.py", line 365, in inspect_pypi_package
    return inspect_pypi_package_sdist(filename, fileobj)
  File "/Users/coleb/binstar/binstar_client/inspect_package/pypi.py", line 244, in inspect_pypi_package_sdist
    release_data = {'version': pop_key(attrs, 'Version'),
  File "/Users/coleb/binstar/binstar_client/inspect_package/uitls.py", line 41, in pop_key
    value = data.pop(k, *d)
KeyError: u'Version'

======================================================================
ERROR: test_sdist_distutils (binstar_client.inspect_package.tests.test_pypi.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/coleb/binstar/binstar_client/inspect_package/tests/test_pypi.py", line 119, in test_sdist_distutils
    package_data, version_data, file_data = pypi.inspect_pypi_package(filename, fd)
  File "/Users/coleb/binstar/binstar_client/inspect_package/pypi.py", line 365, in inspect_pypi_package
    return inspect_pypi_package_sdist(filename, fileobj)
  File "/Users/coleb/binstar/binstar_client/inspect_package/pypi.py", line 244, in inspect_pypi_package_sdist
    release_data = {'version': pop_key(attrs, 'Version'),
  File "/Users/coleb/binstar/binstar_client/inspect_package/uitls.py", line 41, in pop_key
    value = data.pop(k, *d)
KeyError: u'Version'

----------------------------------------------------------------------
Ran 15 tests in 0.478s

Name                                             Stmts   Miss  Cover
---------------------------------------------------------------------
binstar_client/__init__                            226    144    36%
binstar_client/commands/__init__                     0      0   100%
binstar_client/commands/authorizations             137     96    30%
binstar_client/commands/channel                     46     29    37%
binstar_client/commands/config                      66     41    38%
binstar_client/commands/copy                        24     11    54%
binstar_client/commands/groups                      67     50    25%
binstar_client/commands/login                       64     19    70%
binstar_client/commands/logout                      15      6    60%
binstar_client/commands/package                     38     16    58%
binstar_client/commands/remove                      33     22    33%
binstar_client/commands/search                      15      5    67%
binstar_client/commands/show                        66     54    18%
binstar_client/commands/upload                     156     50    68%
binstar_client/commands/whoami                      17      0   100%
binstar_client/errors                               23     18    22%
binstar_client/inspect_package/__init__              0      0   100%
binstar_client/inspect_package/conda                68     23    66%
binstar_client/inspect_package/conda_installer      37     30    19%
binstar_client/inspect_package/ipynb                22      0   100%
binstar_client/inspect_package/pypi                235    119    49%
binstar_client/inspect_package/r                    29     25    14%
binstar_client/inspect_package/uitls                38     12    68%
binstar_client/pprintb                              12      8    33%
binstar_client/requests_ext                         99     43    57%
binstar_client/scripts/__init__                      0      0   100%
binstar_client/scripts/cli                          57     16    72%
binstar_client/utils/__init__                      269    222    17%
binstar_client/utils/detect                         63     35    44%
binstar_client/utils/handlers                       48     13    73%
binstar_client/utils/pprint                         56     40    29%
---------------------------------------------------------------------
Tests: FAILED (errors=3, skipped=2), Coverage: 43%

**Comment by coleb** [comment]
Those stack traces look identical to the ones I get when trying to upload to binstar with the 0.10.3 or 0.10.1 client. Version 0.10.2 is ok.

**Comment by LilaHickey** [comment]
Hi @coleb, we're unfortunately still unable to reproduce your error; can you run a pip freeze and send us the output?

Also, I noticed you're using virtualenv - if possible you might try setting up a conda environment and trying this using python 2.7+.

**Comment by coleb** [comment]
Hi @LilaHickey, so I've been able to track down the root cause of the unicode coercion bug, unfortunately, traveling way back in time. The long and winding path is as follows:

- Originally found the problem on OSX 10.8 with Python 2.7.2 (the default Python interpreter that shipped with OS X 10.8 that we have to support)
- I wasn't able to reproduce the bug on Ubuntu12 and Ubuntu14 with Python 2.7.3 and 2.7.6 respectively
- However, compiling Python 2.7.2 from scratch on RedHat 7, I can reproduce the same test_bdist_egg, test_sdist, and test_sdist_distutils test failures as I'm seeing on OS X 10.8 
- So the bug was fixed by the Python 2.7.3 minor release, specifically, it was fixed by this commit: https://github.com/python/cpython/commit/c13f83ef34befeed284b794560ca0bf050c55112
-- Weirder yet, that commit was to re-apply a change made 5 years earlier: https://github.com/python/cpython/commit/3b3e61ba27cd2ec3eb88315127005dfe987819b1
-- That was reverted one year later here: https://github.com/python/cpython/commit/bc30369136f98913820adbf55c0c3a305ef28345
-- The discussion around putting the functionality back in is here: http://bugs.python.org/issue1548891

So, in summary, Python 2.7.2 (and probably earlier version) have a bug that causes StringIO to silently 'do nothing' if given a unicode string. Need to have Python 2.7.3+. Note, I wasn't able to test Python 2.6 as the clyent library is using some Python 2.7 specific syntax. Python 2.6 is only marginally important for us, we have to support it, but we always have 2.7 around anyway that we can use the binstar client with.

Not sure if you have the ability to target specific Python minor versions with your integration testing. We don't go that insane with our integration testing either. However, we will militantly never upgrade our python interpreters on any of our platforms so that we're always testing "the default python installation".

**Comment by coleb** [comment]
Can you merge this in order to keep compatibility with 2.7.2?

**Comment by LilaHickey** [comment]
Hi @coleb, sorry for the delayed response.  We've been working to reproduce/confirm these errors, and unfortunately your proposed fix causes the unit tests to fail in Python3. I'm reviewing those errors now to see if there is some middle-ground that will satisfy our requirements wall also allowing compatibility with 2.7.2 and will keep you appraised.

**Comment by coleb** [comment]
Thanks for the update, let me know if you need anymore input.

**Comment by LilaHickey** [comment]
@coleb closing this item; your tests should pass on 2.7.2 as of release 0.10.4.  let us know if you're still having a problem with that.
